### PR TITLE
Impute by margin

### PR DESCRIPTION
**Pull request by lgatto**
@cvanderaa - could you review this PR, please.

### CODE REVIEW

File: R/imputation.R [hunk 1]
```diff
@@ -272,53 +303,80 @@ imputeMethods <- function()
...
 ##' @export
 ##' @rdname imputation
-impute_RF <- function(x, ...) {
+impute_RF <- function(x, margin = 2L, ...) {
     requireNamespace("missForest")
-    t(missForest::missForest(t(x), ...)$ximp)
+    margin <- .checkMargin(margin)
+    if (margin == 2L)
+        x <- t(x)
+    res <- missForest::missForest(x, ...)$ximp
```
Comment (cvanderaa) [inline R/imputation.R:376]: I just checked the docs of `missForest`. It assumes:

>The columns correspond to the variables and the rows to the observations.

It then performs random forest in features space (columns for `missForest`, rows for `MsCoreUtils`). So, you should transpose when `margin == 1L`. Not sure whether it makes sense to impute by sample :man_shrugging:

File: R/imputation.R [hunk 2]
```diff
@@ -403,8 +510,44 @@ impute_with <- function(x, val) {
...
     res
-}
\ No newline at end of file
+}
+
+.checkMargin <- function(margin) {
+    margin <- as.integer(margin)
+    if (!margin %in% c(1L, 2L))
+        stop("'margin' must be 1L or 2L")
+    message("Imputing along margin ", margin, ".")
```
Comment (cvanderaa) [inline R/imputation.R:526]: Maybe explicit what margin 1 and 2 mean? eg

```r
message("Imputing along margin ", margin, ifelse(margin == 1L, " (features, rows)", " (samples, columns)"), ".")
```

**Review by cvanderaa — COMMENTED** [review]
Looks great to me, I added a few minor suggestions!

File: R/imputation.R [hunk 3]
```diff
@@ -224,7 +255,7 @@ impute_matrix <- function(x,
                         choices = imputeMethods(),
                         several.ok = FALSE)
     res <- x
-    if (method %in% c("CRILC", "MinDet", "MinProb"))
+    if (method %in% c("CRILC", "MinProb"))
```
[the remark this answers is not in the document]
Comment (lgatto) [reply R/imputation.R:258]: Indeed, fixed.

File: R/imputation.R [hunk 4]
```diff
@@ -272,53 +303,80 @@ imputeMethods <- function()
...
+    if (margin == 2L)
+        return(t(imp_res$data))
+    else imp_res$data
 }
 
 ##' @export
 ##' @rdname imputation
-impute_mle <- function(x, ...) {
+impute_mle <- function(x, margin = 2L, ...) {
     requireNamespace("norm")
```
[the remark this answers is not in the document]
Comment (lgatto) [reply R/imputation.R:336]: Once this PR is merged, I'll send one that adds `impute_mle2()`.

File: R/imputation.R (hunk 1 above)
Comment (lgatto) [reply R/imputation.R:376]: Fixed, thank you for checking.

File: R/imputation.R (hunk 2 above)
Comment (lgatto) [reply R/imputation.R:526]: OK

**Comment by cvanderaa** [comment]
Thanks! I have no further suggestions.

**Comment by lgatto** [comment]
@cvanderaa - I have added the `impute_mle2()` function (from #100) and a new unit test that checks that the dimnames of the imputed matrix aren't modified (which happened when using `norm2`).

**Review by cvanderaa — COMMENTED** [review]
Looks fine to me! :+1:

**Comment by lgatto** [comment]
@sgibb and @jorainer - will merge now but wait a bit more before pushing to Bioc, so feel free to get back with comments later on.

File: R/imputation.R [hunk 5]
```diff
@@ -6,35 +6,63 @@
...
+##' was missed during acquisition (not selected during data dependent
+##' acquisition) or data processing (not identified or with an
+##' identification score below a chosen false discovery threshold). As
+##' such, imputation is by default performed at the *feature
+##' level*. In some cases, such as imputation by zero or a global
+##' minimum value, it doesn't matter. In other cases, it does matter
+##' very much, such as for example when using the minimum value
+##' computed for each margin (i.e. row or column) as in the *MinDet*
+##' method (see below) - do we want to use the minimum of the sample
+##' or of that features? KNN is another such example: do we consider
```
Comment (jorainer) [inline R/imputation.R:54]: "feature" instead of "features"

File: R/imputation.R [hunk 6]
```diff
@@ -50,29 +78,25 @@
...
 ##'   using a deterministic minimal value approach. Considering a
 ##'   expression data with *n* samples and *p* features, for each
 ##'   sample, the missing entries are replaced with a minimal value
 ##'   observed in that sample. The minimal value observed is estimated
 ##'   as being the q-th quantile (default `q = 0.01`) of the observed
-##'   values in that sample. Implemented in the
-##'   `imputeLCMD::impute.MinDet` function. See
-##'   [imputeLCMD::impute.MinDet()] for details and additional
-##'   parameters.
+##'   values in that sample. The implemented in based on the
```
Comment (jorainer) [inline R/imputation.R:98]: "The implemented function is based on ..."

File: R/imputation.R [hunk 7]
```diff
@@ -272,53 +305,95 @@ imputeMethods <- function()
 ##' @param k `numeric(1)` providing the imputation value used for the
 ##'     first and last samples if they contain an `NA`. The default is
 ##'     to use the smallest value in the data.
-impute_neighbour_average <- function(x, k = min(x, na.rm = TRUE)) {
+impute_neighbour_average <- function(x, k = min(x, na.rm = TRUE), margin = 1L) {
```
Comment (jorainer) [inline R/imputation.R:308]: Just a question - would it maybe be better to call that parameter `MARGIN` instead? Such as in `apply`?

**Review by jorainer — COMMENTED** [review]
Looks good to me. Only some typos and a question that can be ignored.

File: R/imputation.R (hunk 7 above)
Comment (lgatto) [reply R/imputation.R:308]: OK, will change it and then push to Bioc.

**Pull request merged**